helpers/kdocker.1
.\" -*- mode: troff; coding: utf-8 -*-
.\" Automatically generated by Pod::Man 5.01 (Pod::Simple 3.43)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" \*(C` and \*(C' are quotes in nroff, nothing in troff, for use with C<>.
.ie n \{\
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\" ========================================================================
.\"
.IX Title "KDOCKER 1"
.TH KDOCKER 1 "16 August, 2024" "Version 5.5" "General Commands Manual"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH NAME
.Vb 1
\& kdocker \- dock any application in the system tray
.Ve
.SH SYNOPSIS
.IX Header "SYNOPSIS"
\&\fBkdocker\fR [\fIoptions\fR]\fR
.SH DESCRIPTION
.IX Header "DESCRIPTION"
.Vb 1
\& KDocker is a program that will dock any application to the system tray
.Ve
.SH OPTIONS
.IX Header "OPTIONS"
.IP \fB\-a\fR 4
.IX Item "-a"
.Vb 1
\& Show author information
.Ve
.IP \fB\-b\fR 4
.IX Item "-b"
.Vb 1
\& Suppress the warning dialog when docking non\-normal windows (blind mode)
.Ve
.IP "\fB\-d\fR \fIsecs\fR" 4
.IX Item "-d secs"
.Vb 2
\& Maximum time in seconds to allow for a command to start and open a window
\& [default: 5 seconds]
.Ve
.IP "\fB\-e\fR \fItype\fR" 4
.IX Item "-e type"
.Vb 6
\& Name matching syntax. Choices are:
\&  n = normal, substring matching [default]
\&  r = regex (regular expression)
\&  u = unix wildcard
\&  w = wildcard
\&  x = W3C XML Schema 1.1
.Ve
.IP \fB\-f\fR 4
.IX Item "-f"
.Vb 1
\& Dock window that has focus (active window)
.Ve
.IP \fB\-h\fR 4
.IX Item "-h"
.Vb 1
\& Display this help, then exit
.Ve
.IP "\fB\-i\fR \fIfile\fR" 4
.IX Item "-i file"
.Vb 1
\& Custom icon path
.Ve
.IP "\fB\-I\fR \fIfile\fR" 4
.IX Item "-I file"
.Vb 2
\& Custom attention icon path. This icon is set if the title
\& of the application window changes while it is iconified.
.Ve
.IP \fB\-j\fR 4
.IX Item "-j"
.Vb 1
\& Case sensitive name (title) matching
.Ve
.IP \fB\-k\fR 4
.IX Item "-k"
.Vb 1
\& Regex minimal matching
.Ve
.IP \fB\-l\fR 4
.IX Item "-l"
.Vb 1
\& Iconify when focus lost
.Ve
.IP \fB\-m\fR 4
.IX Item "-m"
.Vb 1
\& Keep application window showing (don\*(Aqt hide on dock)
.Ve
.IP "\fB\-n\fR \fIname\fR" 4
.IX Item "-n name"
.Vb 1
\& Match window based on its title
.Ve
.IP \fB\-o\fR 4
.IX Item "-o"
.Vb 1
\& Iconify when obscured by other windows
.Ve
.IP "\fB\-p\fR \fIsecs\fR" 4
.IX Item "-p secs"
.Vb 3
\& By default, when the title of the application changes,
\& a popup is displayed from the system tray for 4 seconds
\& Works well with music players
.Ve
.IP \fB\-q\fR 4
.IX Item "-q"
.Vb 1
\& Disable ballooning title changes (quiet)
.Ve
.IP \fB\-r\fR 4
.IX Item "-r"
.Vb 1
\& Remove this application from the pager
.Ve
.IP \fB\-s\fR 4
.IX Item "-s"
.Vb 1
\& Make the window sticky (appears on all desktops)
.Ve
.IP \fB\-t\fR 4
.IX Item "-t"
.Vb 1
\& Remove this application from the taskbar
.Ve
.IP "\fB\-v, \-\-version\fR" 4
.IX Item "-v, --version"
.Vb 1
\& Show the KDocker version string, then exit
.Ve
.IP "\fB\-w\fR \fIwid\fR" 4
.IX Item "-w wid"
.Vb 2
\& Window id of the application to dock
\& Assumes hex number of the form 0x###...
.Ve
.IP "\fB\-x\fR \fIpid\fR" 4
.IX Item "-x pid"
.Vb 2
\& Process id of the application to dock
\& Assumes decimal number of the form ###...
.Ve
.SH AUTHOR
.IX Header "AUTHOR"
.Vb 2
\& KDocker was written by Girish Ramakrishnan
\& Revised and maintained by John Schember
.Ve
.SH "REPORTING BUGS"
.IX Header "REPORTING BUGS"
Bugs and wishes to the KDocker Home Page <https://github.com/user-none/KDocker>


ChangeLog
for version 5.5
    - Better application icon fetching

for version 5.4
    - Avoid deadlock during shut down on UNIX signals

for version 5.3
    - Add configurable attention icon that is shown if iconified app needs attention
    - Fix building with Qt 5.14+
    - Fix for extraneous events generated by KDE 5.14

for version 5.2
    - Add 'Lock to desktop' option
    - Fix bash completion

for version 5.1
    - Support load/save per application settings
    - Various fixes

for version 5.0
    - Support Qt 5.

for version 4.9
    - Assorted bug fixed (mostly minimal).

for version 4.8
    - Format help argument output so that it wraps at a 79 characters.
    - Scrolling (mouse scroll wheel event) on icon can raise and iconify
      window.
    - Add option -e option to specify pattern matching type.
      Accepts n = normal, r = regex, u = unix wildcard, w = wildcard,
      x = xml schema 1.1.
    - Add option -j to enable case senstive matching.
    - Add option -k to set regex minimal matching.

for version 4.7
    - Fix install location of icon referenced by .desktop file so that it
      shows properly.
    - Fix bug where build fails with GCC-4.7.

for version 4.6
    - Fix bug with windows not showing after being restored when KWin is the
      WM.
    - Fix bug with restored windows not gaining focus when KWin is the WM.

for version 4.5
    - Change -n option behavior:
      - -n now acts as -n -y.
      - Name matching now matches any running application. Previously it would
        only match when launching an application.
    - Remove -y option.
    - Change a number of casts from C style to C++ style.

for version 4.4
    - Remove window manager decoration close button can optionally iconify. It
      ended up causing many issues. Fixes:
      - Drag and drop.
      - Moving borderless windows (chrome).
      - Windows moving slightly when undocking.
      - Focus bug (#600045).
    - Remove -c option.
    - Add help text for -i option.
    - Update INSTALL.
    - Default install path is now /usr instead of /usr/local.

for version 4.3
    - Use QtSingleApplication.
    - Window manager decoration close button can optionally iconify.
    - Changed command line option -c: Disables WM close button iconify.
    - Restructure large parts of the appplication.

for version 4.2.2
    - Fix installing translation files.

for version 4.2.1
    - Fix type in .pro file that prevents all files from installing.

for version 4.2
    - Fix disabling balloon title changes menu option.
    - Ignore docked windows and continue looking when matching command by name
      and when matching by specified pid.
    - Bash completion
    - Italian Translation thanks to Alessio Cassibba
    - Change behavior of tray activation. Raise window when not active, and
      iconify when active.
    - Change to the desktop the window was last on when restoring. Only works
      with WMs that support _NET_CURRENT_DESKTOP.
    - Iconify when focus lost.

for version 4.1
    - Fix bug 431297: Fall back to window title and class name when launching if
      pid becomes invalid. Allows commands that detach and get a new pid (rox)
      to dock correctly.
    - New command line option -n: Specify the string to use for fall back when
      launching a command and a pid match cannot be determined.
    - New command line option -y: Force matching command by title and class
      name.
    - New command line option -x: Specify PID of application to dock.
    - Fix a bug that prevented the time out for trying to dock and launched
      command from increasing.
    - Menu item for setting the tray icon.
    - Fix bug: Do not quit when closing docked application from tray menu.

for version 4.0
    - 4.0 rewrite complete.

for version 4.0-preview-1
    - Preview release of Qt 4 port.

for version 1.3
    - Fix bug when reading back balloon timeout on session restoration
    - AutoLaunch enhancement (Request user to provide program name when
      autodetection failed)
    - Indonesian translations from Eko Prasetiyo
    - Italian translations from Alessio Cassibba
    - Updated Brazilian Portuguese and polish translations

for version 1.2
    - Polish translations from KonMan
    - Check WM_NAME to support programs like LimeWire (for lavacube)
    - Store entire application path as part of session
    - Move around command line handling code
    - Configure balloon timeout using "-p" (for roopesh)

for version 1.1
    - German translations from Thomas Runge
    - Dialog detection bug. I replaced all LOG with TRACE (Thanks to Novoselic
      who raised a question which helped me notice it)
    - -v Displays the version
    - -b Suppress the warning when docking non-normal dialogs
      (For John Pettigrew)
    - Fix issue when passing large command line to previous instance 
      (Reported by Arseniy Akopyan)

for version 1.0 CR1
    - Russian translations by Aleksey Kirpichnikov
    - Brazilian Portugese translations by Paulo Ruthes

for version 1.0
    - Spanish translations are back in (Thanks Edulix)
    - Yet another fix for xmms

for version 1.0 rc2
    - "make install" support (with QMake)
    - Multiple desktop awareness (for Sharon)
    - Icons from Marcel Dierkes
    - Use XDND type mechanism to pass arguments to previous instance
    - Spanish translations are out.
    - Czech translations by Jiri Hofman
    - Drag and drop support for Antonio C. Censi
    - Xmms+PlayList fix (for KDE)

for version 1.0 rc1
    - Session Management support (XSMP) for KDE and GNOME
    - AutoLaunch application support that do not support Session Management
    - Wait until system tray comes up on session restoration (Requested by
      Normen)
    - Fix GNOME docked icon resize issue
    - Track tray label appearance and disappearance handling
    - Major code refactoring. QTrayLabel is now a very reusable class

for version 0.9
    - Support to dock the active window (useful with keyboard shortcuts) -f

src/kdocker.cpp
/*
 *  Copyright (C) 2009, 2012 John Schember <john@nachtimwald.com>
 *  Copyright (C) 2004 Girish Ramakrishnan All Rights Reserved.
 *
 * This is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this software; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307,
 * USA.
 */

#include <QCoreApplication>
#include <QTextStream>

#include "constants.h"
#include "kdocker.h"
#include "trayitemmanager.h"

#include <getopt.h>

#include <Xlib.h>


#define ARG_MAX_LEN 30
#define ARG_PRE_PAD 2
#define ARG_POST_PAD 2
#define MAX_HELP_LINE_LEN 79

KDocker::KDocker() {
    m_trayItemManager = new TrayItemManager();
}

KDocker::~KDocker() {
    if (m_trayItemManager) {
        m_trayItemManager->undockAll();
        delete m_trayItemManager;
        m_trayItemManager = 0;
    }
}

void KDocker::undockAll() {
    if (m_trayItemManager) {
        m_trayItemManager->undockAll();
    }
}

void KDocker::cmd(const QStringList &arguments) {
    if (m_trayItemManager) {
        m_trayItemManager->processCommand(arguments);
    }
}

void KDocker::dockAnother() {
    if (m_trayItemManager) {
        m_trayItemManager->selectAndIconify();
    }
}

void KDocker::quit() {
    if (m_trayItemManager) {
        m_trayItemManager->quit();
    }
}

/*
 * handle arguments that output information to the user. We want to handle
 * them here so they are printed on the tty that the application is run from.
 * If we left it up to TrayItemManager with the rest of the arguments these
 * would be printed on the tty the instance was started on not the instance the
 * user is calling from.
 */
void KDocker::preProcessCommand(int argc, char **argv) {
    int option;
    optind = 0; // initialise the getopt static
    while ((option = getopt(argc, argv, Constants::OPTIONSTRING)) != -1) {
        switch (option) {
            case '?':
                printUsage();
                ::exit(1);
                break;
            case 'a':
                printAbout();
                ::exit(0);
                break;
            case 'h':
                printHelp();
                ::exit(0);
                break;
            case 'u':
                printUsage();
                ::exit(0);
                break;
            case 'v':
                printVersion();
                ::exit(0);
                break;
        }
    }
}

// It would be much faster to just format help args appropriately but that would
// require that translators put new lines in the translation. We just run the args
// through this function to format the text properly so they don't have to worry
// about it.
QString KDocker::formatHelpArgs(QList<QPair<QString, QString> > commands) {
    int padding = 0;
    QString out;

    // Find the longest arg below max length.
    for (int i = 0; i < commands.count(); ++i) {
        int length = commands.at(i).first.length();
        if (length > padding && length < ARG_MAX_LEN) {
            padding = length;
        }
    }
    // Pad spaces before and after the longest arg.
    padding += ARG_PRE_PAD + ARG_POST_PAD;

    // Build a formatted string from the arg and its description.
    for (int i = 0; i < commands.count(); ++i) {
        QString arg = commands.at(i).first;
        QString desc = commands.at(i).second;
        
        // Add the arg to the output. If it's over our max arg length
        // start the description on a new line.
        out += QString(' ').repeated(ARG_PRE_PAD);
        if (arg.length() < ARG_MAX_LEN) {
            out += arg.leftJustified(padding - ARG_PRE_PAD, ' ');
        } else {
            out += arg + "\n";
            out += QString(' ').repeated(padding);
        }

        int desc_len = desc.length();
        int last_offset = 0;
        int offset = 0;
        // Move forward in the string then back until we hit a space to break at.
        // If we don't find a space in the segment force a split.
        while (offset < desc_len) {
            if (offset != 0) {
                out += QString(' ').repeated(padding);
            }

            offset += MAX_HELP_LINE_LEN - padding;
            if (offset < desc_len) {
                while (offset > last_offset && desc.at(offset) != ' ') {
                    offset--;
                }
                if (offset == last_offset) {
                    offset += MAX_HELP_LINE_LEN - padding;
                }
            } else {
                offset = desc_len;
            }

            out += desc.mid(last_offset, offset - last_offset) + '\n';
            // Skip any spaces because we're going to start on a new line.
            while (offset < desc_len && desc.at(offset) == ' ') {
                offset++;
            }
            last_offset = offset;
        }
    }

    return out;
}

void KDocker::printAbout() {
    QTextStream out(stdout);

    out << Constants::ABOUT_MESSAGE << Qt::endl;
}

void KDocker::printHelp() {
    QTextStream out(stdout);
    QList<QPair<QString, QString> > commands;

    out << tr("Usage: %1 [options] [command] -- [command options]").arg(qApp->applicationName().toLower()) << Qt::endl;
    out << tr("Docks most applications to the system tray") << Qt::endl;
    out << Qt::endl;
    out << tr("Command") << Qt::endl;
    out << tr("Run command and dock window") << Qt::endl;
    out << tr("Use -- after command to specify options for command") << Qt::endl;
    out << Qt::endl;
    out << tr("Options") << Qt::endl;

    commands.append(qMakePair(QString("-a"),      tr("Show author information")));
    commands.append(qMakePair(QString("-b"),      tr("Don't warn about non-normal windows (blind mode)")));
    commands.append(qMakePair(QString("-d secs"), tr("Maximum time in seconds to allow for command to start and open a window (defaults to 5 sec)")));
    commands.append(qMakePair(QString("-e type"), tr("Name matting syntax. Choices are 'n', 'r', 'w'. n = normal, substring matching (default). r = regex. w = wildcard.")));
    commands.append(qMakePair(QString("-f"),      tr("Dock window that has focus (active window)")));
    commands.append(qMakePair(QString("-h"),      tr("Display this help")));
    commands.append(qMakePair(QString("-i file"), tr("Set the tray icon to file")));
    commands.append(qMakePair(QString("-I file"), tr("Set the attention tray icon to file. This icon is set if the title of the application window changes while it is iconified.")));
    commands.append(qMakePair(QString("-j"),      tr("Case senstive name (title) matching")));
    commands.append(qMakePair(QString("-l"),      tr("Iconify when focus lost")));
    commands.append(qMakePair(QString("-m"),      tr("Keep application window showing (don't hide on dock)")));
    commands.append(qMakePair(QString("-n name"), tr("Match window based on window title")));
    commands.append(qMakePair(QString("-o"),      tr("Iconify when obscured")));
    commands.append(qMakePair(QString("-p secs"), tr("Set ballooning timeout (popup time)")));
    commands.append(qMakePair(QString("-q"),      tr("Disable ballooning title changes (quiet)")));
    commands.append(qMakePair(QString("-r"),      tr("Remove this application from the pager")));
    commands.append(qMakePair(QString("-s"),      tr("Make the window sticky (appears on all desktops)")));
    commands.append(qMakePair(QString("-t"),      tr("Remove this application from the taskbar")));
    commands.append(qMakePair(QString("-v"),      tr("Display version")));
    commands.append(qMakePair(QString("-w wid"),  tr("Window id of the application to dock. Assumes hex number of the form 0x###...")));
    commands.append(qMakePair(QString("-x pid"),  tr("Process id of the application to dock. Assumes decimal number of the form ###...")));
    out << formatHelpArgs(commands);

    out << Qt::endl;
    out << tr("Bugs and wishes to https://github.com/user-none/KDocker") << Qt::endl;
    out << tr("Project information at https://github.com/user-none/KDocker") << Qt::endl;
}

void KDocker::printUsage() {
    QTextStream out(stdout);
    out << tr("Usage: %1 [options] command").arg(qApp->applicationName().toLower()) << Qt::endl;
    out << tr("Try `%1 -h' for more information").arg(qApp->applicationName().toLower()) << Qt::endl;
}

void KDocker::printVersion() {
    QTextStream out(stdout);
    out << tr("%1 version: %2").arg(qApp->applicationName()).arg(qApp->applicationVersion()) << Qt::endl;
    out << tr("Using Qt version: %1").arg(qVersion()) << Qt::endl;
}
